### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by skwong25**

### CODE REVIEW

**Comment by skwong25** [comment]
Style changes still to happen re. spaces after functions. 
ReadMe still to happen. 
For consistency, BreedRepository should really also be in a class and have its own unit tests. 
Still some doubling up of code that needs to be DRY-ed.
Postman issues are still unresolved for any requests that pass in objects (GET/DEL/PUT by id, POST) please lets have a look at this together. Would be good to know what is going wrong.

File: src/buildServer.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// Builds Server with the non-test Id Generator
```
Comment (fractalwrench) [inline src/buildServer.js:1]: looks good 👍

File: src/catsDb.js [hunk 2]
```diff
@@ -122,6 +126,10 @@ class CatRepository {
...
 }
 
-module.exports = new CatRepository();
+module.exports.catClass = CatClass
+// export a class by attaching it as a property of the module.exports object 
 
 
+// Note that if a module is imported multiple times, but with the same specifier (i.e. path), 
+// the JavaScript specification guarantees that you’ll receive the same module instance. 
+// Above, we should expect that although the class instance is imported in app.js AND catsRouter, it will be the same instance.
```
Comment (fractalwrench) [inline src/catsDb.js:135]: great comment :)

File: src/catsDb.js [hunk 3]
```diff
@@ -77,14 +81,14 @@ class CatRepository {
         if (catIndex >= 0) {
             return catIndex; 
         } else {
-            return null; 
+            return null; // TODO - check if we can DRY this check for id. We are passing a lot of 'nulls' around. catIndex is either an index or -1.
```
Comment (fractalwrench) [inline src/catsDb.js:84]: I'd go further and say that you could alter all the places that currently check null to check if the index is -1 instead. By convention if an index is -1 in programming it means that the element is not present in an array.

File: src/catsDb.js [hunk 4]
```diff
@@ -1,34 +1,38 @@
-// class 'CatRepository' serves as a template for creating new cat objects with props & methods
+// class 'CatClass' serves as a template for creating new cat objects with props & methods
+// example of dependancy injection:
 
-class CatRepository {
-    constructor() {
-        this.cats = [ // array of objects - if object, change to >> this.cats:{ 1:{}, 2:{}, 3:{}}
-            {
-                id: 1,
+class CatClass {
```
Comment (fractalwrench) [inline src/catsDb.js:4]: The requirements say that this should be named `CatRepository`. This is a more specific and descriptive name - if you were looking at this codebase for the first time you wouldn't know what a `CatClass` is, but you'd have an inkling of what a repository is and what it does.

File: test/apiTest.js [hunk 5]
```diff
@@ -150,36 +179,37 @@ describe('POST /cats', function () {
...
 })
+// Error: expected 400 "Bad Request", got 500 "Internal Server Error"
 
 describe('GET /breeds', function () {
     it('respond with json object containing a list of all breeds', function (done) {
-        request(app)
+        request(appTest)
             .get('/breeds')
-            .set('Accept', 'application/json')
+            .set('Accept', 'appTestlication/json')
```
Comment (fractalwrench) [inline test/apiTest.js:196]: there are a couple of similar typos in the tests

File: test/apiTest.js [hunk 6]
```diff
@@ -1,58 +1,79 @@
...
 
 describe('GET /cats', function () {
-    it('responds with json object containing a list of cat ids and name only', function (done) {
-        request(app)
+    it('responds with json object containing a list of cats with id & name properties only', function (done) {
+        request(appTest)
             .get('/cats')
             .set('Accept', 'application/json')
             .expect('Content-Type', /json/)
+            .expect(function (res) {
```
Comment (fractalwrench) [inline test/apiTest.js:25]: it'd be good to clean up this area - there seem to be a lot of comments which might not be fully resolved yet?

File: test/apiTest.js [hunk 7]
```diff
@@ -1,58 +1,79 @@
...
+
+describe('GET /cats/:id', function () {
+    it('respond with 404 cat id not found - passes the validId check', function (done) {
+        request(appTest)
+            .get('/cats/oc12C0X3-g')
             .set('Accept', "text/html; charset=utf-8")
             .expect('Content-Type', "text/html; charset=utf-8") 
             .expect(404) 
-            .expect(`cat id '9' not found in database`) 
+            .expect(`cat id 'oc12C0X3-g' not found in database`)
```
Comment (fractalwrench) [inline test/apiTest.js:55]: These tests are great 👌

File: src/app.js [hunk 8]
```diff
@@ -1,49 +1,66 @@
 // CAT API
 
-const express = require('express');         // imports express lib module
-const app = express();                     // express() instantiates app 
-const morgan = require('morgan');
-const bodyParser = require('body-parser')
+// QUESTION: Am I confused re. repositories?
```
Comment (fractalwrench) [inline src/app.js:3]: are you still confused?

File: src/app.js [hunk 9]
```diff
@@ -1,49 +1,66 @@
...
+    app.use('/breeds', breedsRouter);
 
+    const PORT = 4001;
+
+    app.use(morgan('tiny'));
+
+    // error-handling sends back error responses 
+    app.use((err, req, res, next) => {
+        let status = err.status || 500  
+        let message = err.message || 'test message'
```
Comment (fractalwrench) [inline src/app.js:44]: it's good you're thinking about fallback error messages - maybe you could replace the string with something like 'an unexpected error has occurred'?

File: test/apiTest.js [hunk 10]
```diff
@@ -1,58 +1,79 @@
 // next we will write an API test for '/breed routes' GET and GET by id:
```
Comment (fractalwrench) [inline test/apiTest.js:1]: this is good :) you will probably want to split that up into a separate test file when you do that, but that should be done in a separate PR.

File: src/breedsDb.js [hunk 11]
```diff
@@ -18,13 +18,63 @@ class BreedRepository {
...
+    set setBreeds(breedsCopy) {
+        this.breeds = breedsCopy;
+    }
+
+    findIndexById(id) {
+        const breedsCopy = this.getAllBreeds; 
+        const foundBreed = breedsCopy.findIndex((breed) => {
+            return breed.breedId === id;       
+        });
+        return foundBreed;  // foundBreed is either -1 or a valid index
```
Comment (fractalwrench) [inline src/breedsDb.js:30]: at first I thought this was a boolean variable but it's not - perhaps consider renaming

File: src/breedsDb.js [hunk 12]
```diff
@@ -18,13 +18,63 @@ class BreedRepository {
...
         const breedsCopy = this.getAllBreeds; 
-        const idMatchArray = breedsCopy.filter((breed) => { 
+        const idMatchArray = breedsCopy.filter((breed) => {  // could use findIndexById instead of array.filter
             return breed['breedId'] == id; 
         })
         return idMatchArray[0];
     }
+
+    updateBreedById (id, object) {
+        const breedsCopy = this.getAllBreeds;
```
Comment (fractalwrench) [inline src/breedsDb.js:42]: I think this might be a bug, as it's not invoking the function? Or perhaps my JS syntax is off..

File: src/breedsDb.js [hunk 13]
```diff
@@ -18,13 +18,63 @@ class BreedRepository {
...
+
+    updateBreedById (id, object) {
+        const breedsCopy = this.getAllBreeds; 
+        const index = this.findIndexById(id); // 
+        if (index === -1) {
+            console.log(`index not present: ${index}`);
+            return null;
+        } else {
+            for (let key in object) {
+                if (breedsCopy[index][key] !== object[key]) {
```
Comment (fractalwrench) [inline src/breedsDb.js:49]: style nit: consider creating local variables for `breedsCopy[index]` and `object[key]`, as this will reduce the mental overhead required to read the code.

File: src/breedsRouter.js [hunk 14]
```diff
@@ -1,16 +1,26 @@
+// TODO provide DEL, POST, PUT routes also.
```
Comment (fractalwrench) [inline src/breedsRouter.js:1]: That can be part of a separate PR - let's get the current functionality squared away first.

File: src/breedsRouter.js [hunk 15]
```diff
@@ -1,16 +1,26 @@
+// TODO provide DEL, POST, PUT routes also. 
 
 const express = require('express');  
 const breedsRouter = express.Router();
 
-const BreedRepository = require('./breedsDb.js');
+const breedRepository = require('./breedsDb.js');
+
+
+const isInvalidString = (value) => {
```
Comment (fractalwrench) [inline src/breedsRouter.js:9]: These seem familiar! could you consider extracting them into a separate file from catsRouter/breedsRouter to avoid duplication? You could also consider unit tests for the individual functions if they're not covered by tests already, as it should be pretty straightforward to test these in isolation from other components.

File: src/breedsRouter.js [hunk 16]
```diff
@@ -19,21 +29,95 @@ const isBreedId = (req, res, next) => {
...
+    req.object = objToCheck;            
+    console.log('object values checked');     
+    next();
+} // TODO Review if we can use same code as in catsRouter with diff parameters? Maybenot
+
 // GET route all 
-breedsRouter.get('', (req, res, next) => {
-    const breeds = BreedRepository.getAllBreeds; // this returns an array of objects
+breedsRouter.get('/', (req, res, next) => {
+    const breeds = breedRepository.getAllBreeds; // this returns an array of objects
```
Comment (fractalwrench) [inline src/breedsRouter.js:71]: this rename is good and matches JS conventions 👍

File: src/catsRouter.js [hunk 17]
```diff
@@ -1,164 +1,169 @@
...
-    if (isUpdated) {
-        console.log(`cat id '${req.id}' successfully updated`)
-        res.send(isUpdated); // may need to use: res.send(JSON.stringify(updatedCat)); 
-    } else {
-        const newError = new Error(`cat id '${req.id}' not found in database`);
-        newError.status = 404; 
-        next(newError); 
+    // checks object data format
+    checkObjFormat (req, res, next) { 
+        const objectConstructor = ({}).constructor; // in strict mode, 'this' in functions is undefined.
```
Comment (fractalwrench) [inline src/catsRouter.js:45]: it'd be great if you could link to a blog post or something on this - often this can be a helpful reminder when reading the code again.

File: src/catsRouter.js [hunk 18]
```diff
@@ -1,164 +1,169 @@
...
+            console.log(`Property '${invalidKey[0]}' is invalid`);
+            const newError = new Error(`Property '${invalidKey[0]}' is invalid`);
+            newError.status = 400;
+            return next(newError);
+        } else {
+            console.log('object keys checked'); 
+            req.object = objToCheck; 
+            next();
+        }
+    } // FIXME this seems convoluted just to check if an object key is valid
```
Comment (fractalwrench) [inline src/catsRouter.js:89]: There are probably a few things which could be altered here

**Review by fractalwrench — COMMENTED** [review]
Looking good :) I took a first pass review, and will try and test it out a bit more tomorrow.

File: test/apiTest.js (hunk 5 above)
Comment (skwong25) [reply test/apiTest.js:196]: haha my first dabblings in the Find & Replace tool.

File: src/catsRouter.js (hunk 17 above)
Comment (skwong25) [reply src/catsRouter.js:45]: I keep all useful blog posts links in my own notes to avoid cluttering up the code: 
https://www.vojtechruzicka.com/javascript-this-keyword/

File: src/catsRouter.js (hunk 18 above)
Comment (skwong25) [reply src/catsRouter.js:89]: Well duh

File: src/catsRouter.js (hunk 17 above)
Comment (fractalwrench) [reply src/catsRouter.js]: It's good you're keeping notes - but in this particular case it'd probably be helpful to link off to documentation about what is happening in case there are any JS newbies (like me) reading the code.

File: src/catsDb.js (hunk 3 above)
Comment (skwong25) [reply src/catsDb.js]: Done.

File: src/catsDb.js (hunk 4 above)
Comment (skwong25) [reply src/catsDb.js]: Done.

File: src/catsRouter.js (hunk 18 above)
Comment (skwong25) [reply src/catsRouter.js]: Excue my flippancy :) Let me know if I got all the 'things' in my latest PR.

File: src/catsDb.js (hunk 4 above)
Comment (fractalwrench) [reply src/catsDb.js]: This can just be named `CatRepository` - adding class onto the end is nearly always omitted as it's possible to get that information from the context. Although I can appreciate this naming might help a little when starting out it'd be best to stick with convention

File: src/breedsRouter.js [hunk 19]
```diff
@@ -19,21 +29,95 @@ const isBreedId = (req, res, next) => {
     };
 }
 
+// checks keys are valid 
+const checkBreedKeys = (req, res, next) => {
+    const breedObjToCheck = req.body                        // bodyParser auto attaches JSON object to req.body
+    const validKeys = ["breedId", "name", "description"];
+    const arrayOfKeys = Object.keys(breedObjToCheck);
+
+    for (let i = 0; i<arrayOfKeys; i++) {
```
Comment (fractalwrench) [inline src/breedsRouter.js:38]: style nit: missing whitespace in the for loop

**Review by fractalwrench — APPROVED** [review]
Let's focus on resolving the outstanding comments then get this merged. It'd be really good to try and aim for [small PRs](https://medium.com/@hugooodias/the-anatomy-of-a-perfect-pull-request-567382bb6067) in future as that'll make it a bit easier to review, and hopefully get things merged quicker. It's also good practice for you as you'll get to practice how to break functionality into logical chunks :)

File: src/catsDb.js (hunk 4 above)
Comment (skwong25) [reply src/catsDb.js]: Its the distinguish easier the class from the class instance, but yes ok.
